src/lib.rs

```diff
@@ -608,138 +808,328 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> HotShot<TYPES::ConsensusType
+        // replicas should only send votes or if they timed out, timeouts
+        match msg.0 {
+            Left(general_message) => match general_message {
+                GeneralConsensusMessage::Proposal(_)
+                | GeneralConsensusMessage::InternalTrigger(_) => {
+                    warn!("Received a direct message for a proposal. This shouldn't be possible.");
+                }
+                // this is ONLY intended for next leader
+                c @ GeneralConsensusMessage::Vote(_) => {
+                    let channel_map = self.next_leader_channel_map.upgradable_read().await;
+
+                    // check if
+                    // - is in fact, actually is the next leader
+                    // - the message is not stale
+                    let is_leader = self
+                        .inner
+                        .clone()
+                        .exchanges
+                        .quorum_exchange()
+                        .is_leader(msg_time + 1);
+                    if !is_leader || msg_time < channel_map.cur_view {
+                        warn!(
+                            "Throwing away VoteType<TYPES>message for view number: {:?}",
+                            msg_time
+                        );
+                        return;
+                    }
+
+                    let chan = Self::create_or_obtain_chan_from_read(msg_time, channel_map).await;
+
+                    if chan
+                        .sender_chan
+                        .send(Left(ProcessedGeneralConsensusMessage::new(c, sender)))
+                        .await
+                        .is_err()
+                    {
+                        error!("Failed to send to next leader!");
+                    }
+                }
+                GeneralConsensusMessage::ViewSync(_) => todo!(),
+            },
+            Right(committee_message) => {
+                match committee_message {
+                    c @ CommitteeConsensusMessage::DAVote(_) => {
+                        let channel_map = self.da_leader_channel_map.upgradable_read().await;
+
+                        // check if
+                        // - is in fact, actually is the next leader
+                        // - the message is not stale
+                        let is_leader = self
+                            .inner
+                            .clone()
+                            .exchanges
+                            .committee_exchange()
+                            .is_leader(msg_time);
+                        if !is_leader || msg_time < channel_map.cur_view {
+                            warn!(
+                                "Throwing away VoteType<TYPES>message for view number: {:?}, Channel cur view: {:?}",
+                                msg_time,
+                                channel_map.cur_view,
+                            );
+                            return;
+                        }
```

Comment:
It might be worth using [this](https://doc.rust-lang.org/std/any/fn.type_name.html) to print the type name instead of `VoteType<TYPES>`. Maybe post monomorphized types would be more informative? Or maybe printing both?